node/src/db/models/consensus/instructions.rs

```diff
@@ -0,0 +1,275 @@
+use crate::{
+    db::{
+        models::{InstructionStatus, NewAssetStateAppendOnly, NewTokenStateAppendOnly},
+        utils::errors::DBError,
+    },
+    types::{AssetID, InstructionID, NodeID, ProposalID, TemplateID, TokenID},
+};
+use chrono::{DateTime, Utc};
+use deadpool_postgres::Client;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tokio_pg_mapper::{FromTokioPostgresRow, PostgresMapper};
+use tokio_postgres::types::Type;
+
+#[derive(Clone, Deserialize, Serialize, PostgresMapper, PartialEq, Debug)]
+#[pg_mapper(table = "instructions")]
+pub struct Instruction {
+    pub id: InstructionID,
+    pub initiating_node_id: NodeID,
+    pub signature: String,
+    pub asset_id: AssetID,
+    pub token_id: Option<TokenID>,
+    pub template_id: TemplateID,
+    pub contract_name: String,
+    pub status: InstructionStatus,
+    pub params: Value,
+    pub created_at: DateTime<Utc>,
+    pub updated_at: DateTime<Utc>,
+    pub proposal_id: Option<ProposalID>,
+}
+
+/// Query parameters for adding new instruction record
+#[derive(Default, Clone, Debug)]
+pub struct NewInstruction {
+    pub id: InstructionID,
+    pub initiating_node_id: NodeID,
+    pub signature: String,
+    pub asset_id: AssetID,
+    pub token_id: Option<TokenID>,
+    pub template_id: TemplateID,
+    pub contract_name: String,
+    pub status: InstructionStatus,
+    pub params: Value,
+}
+
+/// Query parameters for optionally updating instruction fields
+#[derive(Default, Clone, Debug)]
+pub struct UpdateInstruction {
+    pub status: Option<InstructionStatus>,
+    pub proposal_id: Option<ProposalID>,
+}
+
+impl Instruction {
+    pub async fn find_pending(client: &Client) -> Result<Option<(AssetID, Vec<Self>)>, DBError> {
+        let stmt = "
+            SELECT i.*
+            FROM instructions i
+            JOIN (
+                SELECT i.asset_id
+                FROM instructions i
+                JOIN asset_states ast ON ast.asset_id = i.asset_id
+                WHERE i.status = 'Pending'
+                AND ast.blocked_until <= now()
+                LIMIT 1
+            ) i2 ON i.asset_id = i2.asset_id
+            AND i.status = 'Pending'
+        ";
```

Comment:
Would not that pick previous asset states for same asset which have blocked_until <= now() ?We would nee to rethink this one

Reply:
Actually I was wrong, asset_id is UNIQUE, I keep confusing myself that there might be few AssetStates for given asset. So this should be good!

Reply:
I will also add Processing state, which precedes Pending, that will respond for instructions which are being executed.